runner/tasks.py

```diff
@@ -0,0 +1,211 @@
+from invoke import task
+import os
+import datetime
+import re
+import logging
+import requests
+import csv
+
+
+logging.basicConfig(level=logging.INFO)
+
+
+def _get_remaining_tests(ctx):
+    selectors = [c["selector"] for c in ctx.categories if "selector" in c]
+
+    return " and ".join([f"not {s}" for s in selectors])
+
+
+def _init_output_dir(ctx, test_name):
+    if "output_dir" not in ctx:
+        dt = datetime.datetime.now().isoformat()
+        dir_name = f"{ctx.base_output_dir}/{test_name}_{dt}"
+        os.makedirs(dir_name)
+        ctx.output_dir = dir_name
+    return ctx.output_dir
+
+
+def _pretty_print(dicts, keys):
+    """
+    pretty print an array of dict as a table
+
+    shamelessly taken and adapted from geocoder-tester
+    """
+    if not dicts:
+        return []
+    # Compute max length for each column.
+    lengths = {}
+    for key in keys:
+        lengths[key] = len(key) + 2  # Surrounding spaces.
+    for d in dicts:
+        for key in keys:
+            i = len(str(d.get(key, "")))
+            if i > lengths[key]:
+                lengths[key] = i + 2  # Surrounding spaces.
+    out = [""]
+    cell = "{{{key}:^{length}}}"
+    tpl = "|".join(cell.format(key=key, length=lengths[key]) for key in keys)
+    # Headers.
+    out.append(tpl.format(**dict(zip(keys, keys))))
+    # Separators line.
+    out.append(tpl.format(**dict(zip(keys, ["—" * lengths[k] for k in keys]))))
+    for d in dicts:
+        row = {}
+        l = lengths.copy()
+        for key in keys:
+            value = d.get(key) or "_"
+            row[key] = value
+        # Recompute tpl with lengths adapted to failed rows (and thus ansi
+        # extra chars).
+        tpl = "|".join(cell.format(key=key, length=l[key]) for key in keys)
+        out.append(tpl.format(**row))
+    return out
+
+
+RESULT_LINE_PATTERN = re.compile("===+.*===+")
+
+RESULT_PATTERN = re.compile(
+    "===+( (?P<failed>\d+) failed,)? ((?P<success>\d+) passed)?.*in (?P<time>.*) seconds.*"
```

Comment:
Very minor note: this still does not match the number of failures in
`========================= 952 failed in 17.06 seconds ==========================`
(the comma is absent)